### ISSUE TITLE
False negative: Context manager which suppresses errors doesn't trigger `[possibly-undefined]`

### ISSUE THREAD

**Issue by bzoracler** [python/mypy#17664]
**Bug Report, To Reproduce, & Actual Behaviour**

The following snippet will fail at runtime, which mypy should catch with `[possibly-undefined]` (see [mypy Playground](https://mypy-play.net/?mypy=latest&python=3.12&gist=ffd0896ef508012a9825872ae6df3778)):

```python
# mypy: enable-error-code=possibly-undefined

from typing import Literal

class ExceptionSuppressor:
    def __enter__(self) -> None:
        return

    def __exit__(self, *args: object, **kwargs: object) -> Literal[True]:
        return True

def get_num() -> int:
    raise RuntimeError

with ExceptionSuppressor():
    var = get_num()

print(var)  # Runtime failure
```

**Expected Behavior**

```python
print(var)  # mypy: Name "var" may be undefined [possibly-undefined]
```

**Your Environment**

- Mypy version used: 1.10.1, 1.11.1
- Mypy command-line flags: `enable-error-code=possibly-undefined`
- Mypy configuration options from `mypy.ini` (and other config files): None
- Python version used: 3.10, 3.12

**Comment by sterliakov** [comment]
I don't think this is feasible for any static analysis tool. To emit [possibly-undefined] here, `mypy` needs to understand that `ExceptionSuppressor.__exit__` always returns True, swallowing the exception. This is only doable in most simple cases where the return type can be statically inferred to `Literal[True]` and requires more special casing later to detect the "this assignment may not happen silently" scenario. Far beyond reasonable IMO, especially given how rare this case is.

**Comment by bzoracler** [comment]
I had a brief look, and the largest changes required to support this seem to be treating `with` statements as potential branch statements, rather than needing to spend effort handling the return type of `__exit__`.
* Analysis of `__exit__`'s return for error-suppression is [already done in the type-checking phase](https://github.com/python/mypy/blob/5e119d0062ca425d927c0f036a63d95bf0cad367/mypy/checker.py#L5221)
* `possibly-undefined` is [analysed *after* the type-checking phases](https://github.com/python/mypy/blob/5e119d0062ca425d927c0f036a63d95bf0cad367/mypy/build.py#L3436-L3448)
* I don't think it's recommended to return anything either than `bool | None` or some subtype thereof in `__exit__`, based on [these typeshed guidelines](https://github.com/python/typeshed/blob/main/CONTRIBUTING.md#context-managers). In any case, I would expect it to be rare to return an arbitrary object (as you can't assign the return value of `__exit__` in runtime code using the context manager `with` statement syntax), but mypy [already exposes](https://github.com/python/mypy/blob/5e119d0062ca425d927c0f036a63d95bf0cad367/mypy/checker.py#L5786-L5800) basic logic to statically infer `bool(<arbitrary expression>)` as `True` regardless.

I won't comment on the rarity of needing this - my use case was trying to capture and log errors when interfacing with an asynchronous, message-driven third-party program, where the exceptions can all be suppressed (they were all recoverable back to normal execution), but I didn't necessarily know what those exceptions would be, and I didn't feel like littering verbose `try...except builtins.Exception`s all over the place.
